src/main/scala/is/hail/io/bgen/LoadBgen.scala

```diff
@@ -30,54 +27,73 @@ case class BgenHeader(
-      val reader = new HadoopFSDataBinaryReader(is)
-      reader.seek(0)
-
-      for (i <- 1 to bState.nVariants) {
-        reader.seek(position)
-
-        if (bState.version == 1)
-          reader.readInt() // nRows for v1.1 only
-
-        val snpid = reader.readLengthAndString(2)
-        val rsid = reader.readLengthAndString(2)
-        val chr = reader.readLengthAndString(2)
-        val pos = reader.readInt()
-
-        val nAlleles = if (bState.version == 2) reader.readShort() else 2
-        assert(nAlleles >= 2, s"Number of alleles must be greater than or equal to 2. Found $nAlleles alleles for variant '$snpid'")
-        (0 until nAlleles).foreach { i => reader.readLengthAndString(4) }
-
-        position = bState.version match {
-          case 1 =>
-            if (bState.compressed)
-              reader.readInt() + reader.getPosition
-            else
-              reader.getPosition + 6 * bState.nSamples
-          case 2 =>
-            reader.readInt() + reader.getPosition
-        }
-
-        dataBlockStarts(i) = position
-      }
+    val typedRowFields = Array(
+      "locus" -> TLocus.schemaFromRG(referenceGenome),
+      "alleles" -> TArray(TString()),
+      "offset" -> TInt64())
+
+    val requestedType: MatrixType = MatrixType.fromParts(
+      globalType = TStruct.empty(),
+      colKey = Array("s"),
+      colType = TStruct("s" -> TString()),
+      rowType = TStruct(typedRowFields: _*),
+      rowKey = Array("locus", "alleles"),
+      rowPartitionKey = Array("locus"),
+      entryType = TStruct())
+
+    val attributes = Map("reference_genome" -> rg.orNull,
+      "contig_recoding" -> Option(contigRecoding).getOrElse(Map.empty[String, String]),
+      "skip_invalid_loci" -> skipInvalidLoci)
+
+    val rowType = requestedType.rowType
+    val offsetIdx = rowType.fieldIdx("offset")
+    val (keyType, kf) = rowType.select(Array("locus", "alleles"))
+
+    val rvds = files.map { f =>
+      val reader = MatrixBGENReader(
+        Array(f), None, Some(1), None,
+        rg,
+        Option(contigRecoding).getOrElse(Map.empty[String, String]),
+        skipInvalidLoci,
+        Map.empty[String, Seq[Int]],
+        createIndex = true)
+
+      val mt = new MatrixTable(hc, MatrixRead(
+        requestedType,
+        dropCols = true,
+        dropRows = false,
+        reader))
+
+      assert(mt.nPartitions == 1)
+      mt.rvd
     }
 
-    IndexBTree.write(dataBlockStarts, indexFile, hConf)
+    val unionRVD = if (files.length == 1) rvds.head else RVD.union(rvds)
+    assert(unionRVD.getNumPartitions == files.length)
 
+    unionRVD.mapPartitionsWithIndex({ (i, it) =>
+      val iw = new IndexWriter(hConf, files(i) + ".idx", keyType, TStruct(required = true), attributes = attributes)
+      it.foreach { rv =>
+        val r = UnsafeRow.readBaseStruct(rowType, rv.region, rv.offset)
+        iw += (kf(r), r.getLong(offsetIdx), Row())
+        rv
+      }
+      iw.close()
+      Iterator.single(1)
```

Comment:
`Iterator.empty()` is fine since we ignore the result anyway

Reply:
Iterator.empty throws an exception:

```
E           hail.utils.java.FatalError: ArrayStoreException: [Ljava.lang.Object;
E
E           Java stack trace:
E           org.apache.spark.SparkDriverExecutionException: Execution error
E           	at org.apache.spark.scheduler.DAGScheduler.handleTaskCompletion(DAGScheduler.scala:1186)
E           	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1711)
E           	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
E           	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
E           	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
E           	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
E           	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
E           	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2043)
E           	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2062)
E           	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2087)
E           	at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:936)
E           	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
E           	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
E           	at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
E           	at org.apache.spark.rdd.RDD.collect(RDD.scala:935)
E           	at is.hail.io.bgen.LoadBgen$.index(LoadBgen.scala:93)
E           	at is.hail.HailContext.indexBgen(HailContext.scala:359)
```